faiss_vector_cache.go

```diff
@@ -18,92 +18,130 @@
 package zap
 
 import (
+	"encoding/binary"
 	"sync"
 	"sync/atomic"
 	"time"
 
+	"github.com/RoaringBitmap/roaring"
 	faiss "github.com/blevesearch/go-faiss"
 )
 
-type vectorIndexCache struct {
-	closeCh chan struct{}
-	m       sync.RWMutex
-	cache   map[uint16]*cacheEntry
-}
-
-type ewma struct {
-	alpha float64
-	avg   float64
-	// every hit to the cache entry is recorded as part of a sample
-	// which will be used to calculate the average in the next cycle of average
-	// computation (which is average traffic for the field till now). this is
-	// used to track the per second hits to the cache entries.
-	sample uint64
-}
-
-type cacheEntry struct {
-	tracker *ewma
-	// this is used to track the live references to the cache entry,
-	// such that while we do a cleanup() and we see that the avg is below a
-	// threshold we close/cleanup only if the live refs to the cache entry is 0.
-	refs  int64
-	index *faiss.IndexImpl
-}
-
 func newVectorIndexCache() *vectorIndexCache {
 	return &vectorIndexCache{
 		cache:   make(map[uint16]*cacheEntry),
 		closeCh: make(chan struct{}),
 	}
 }
 
+type vectorIndexCache struct {
+	closeCh chan struct{}
+	m       sync.RWMutex
+	cache   map[uint16]*cacheEntry
+}
+
 func (vc *vectorIndexCache) Clear() {
 	vc.m.Lock()
 	close(vc.closeCh)
 
 	// forcing a close on all indexes to avoid memory leaks.
 	for _, entry := range vc.cache {
-		entry.closeIndex()
+		entry.close()
 	}
 	vc.cache = nil
 	vc.m.Unlock()
 }
 
-func (vc *vectorIndexCache) loadVectorIndex(fieldID uint16,
-	indexBytes []byte) (vecIndex *faiss.IndexImpl, err error) {
-	cachedIndex, present := vc.isIndexCached(fieldID)
-	if present {
-		vecIndex = cachedIndex
-		vc.incHit(fieldID)
+func (vc *vectorIndexCache) loadOrCreate(fieldID uint16, mem []byte, except *roaring.Bitmap) (
+	index *faiss.IndexImpl, vecDocIDMap map[int64]uint32, vecIDsToExclude []int64, err error) {
+	entry := vc.fetch(fieldID)
+	if entry != nil {
+		index, vecDocIDMap = entry.load()
+		if except != nil && !except.IsEmpty() {
+			for vecID, docID := range vecDocIDMap {
+				if except.Contains(docID) {
+					vecIDsToExclude = append(vecIDsToExclude, vecID)
+				}
+			}
+		}
 	} else {
-		vecIndex, err = vc.createAndCacheVectorIndex(fieldID, indexBytes)
+		index, vecDocIDMap, vecIDsToExclude, err = vc.createAndCache(fieldID, mem, except)
+	}
+	return index, vecDocIDMap, vecIDsToExclude, err
+}
+
+func (vc *vectorIndexCache) fetch(fieldID uint16) *cacheEntry {
+	vc.m.RLock()
+	defer vc.m.RUnlock()
+
+	entry, ok := vc.cache[fieldID]
+	if !ok {
+		return nil
 	}
-	vc.addRef(fieldID)
-	return vecIndex, err
+
+	entry.addRef()
```

Comment:
I see that an entry.addRef() is followed immediately entry.load(). for better readability maybe we can move the addRef() invocation inside the load() function and although it might mean have a RLock() there i dont think it would cause functional defects. Please think upon the same and see if you think it's worth doing it

Reply:
It might be better to have the addRef with load. I don't think the locks affect anything here, but I'll check once before making this change